File: mandelbrot/TypeScript/src/renderer.ts
import { exit } from "process";
import { Complex } from "./complex";

export function render(
  pixels: number[],
  bounds: [number, number],
  upperLeft: Complex,
  lowerRight: Complex) {

  if (pixels.length !== bounds[0] * bounds[1]) {
    console.log("Wrong size of pixels array");
    exit(1);
  }

  const width = lowerRight.re - upperLeft.re;
  const height = upperLeft.im - lowerRight.im;

  for(let row = 0; row < bounds[1]; row++) {
    for(let column = 0; column < bounds[0]; column++) {

      const re = upperLeft.re + column * width / bounds[0];
      const im = upperLeft.im - row * height / bounds[1];
      const point = new Complex(re, im);

      let et = escapeTime(point, 255);
      pixels[row * bounds[0] + column] = (et === null) ? 0 : 255 +- et;
    }
  }
}

function escapeTime(c: Complex, limit: number): number | null {
  let zr = 0.0, zi = 0.0;
  for (let it = 0; it < limit; it++) {
    const zr2 = zr * zr;
    const zi2 = zi * zi;

    if (zr2 + zi2 > 4.0) {
      return it;
    }

    const temp = zr2 - zi2 + c.re;
    zi = 2.0 * zr * zi + c.im;
    zr = temp;
  }

  return null;
}

// function escapeTime(c: Complex, limit: number): number | null {
//   let z = new Complex(0.0, 0.0);
//   for(let it = 0; it < limit; it++) {
//     if (z.normSqr() > 4.0) {
//       return it;
//     }

//     z = z.multiply(z).add(c);
//   }

//   return null;
// }

// function pixelToPoint(
//   bounds: [number, number],
//   pixel: [number, number],
//   upperLeft: Complex,
//   lowerRight: Complex): Complex {
//   let [width, height] = [lowerRight.re - upperLeft.re, upperLeft.im - lowerRight.im];
//   return new Complex(
//     upperLeft.re + pixel[0] * width / bounds[0],
//     upperLeft.im - pixel[1] * height / bounds[1]
//   );
// }

File: mandelbrot/C#/Renderer.cs
using System.IO.Compression;
using System.Numerics;

namespace MandelSharp;

public class Renderer
{
  private const int MaxIterations = 255;
  private const double EscapeRadius = 2.0;

  public static void Render(MandelbrotConfig config, int[] pixels)
  {
    var (pixelsWidth, pixelsHeight) = config.Pixels;

    double width = config.LowerRight.Real - config.UpperLeft.Real;
    double height = config.UpperLeft.Imaginary - config.LowerRight.Imaginary;

    for (int row = 0; row < pixelsHeight; row++)
    {
      for (int column = 0; column < pixelsWidth; column++)
      {
        var re = config.UpperLeft.Real + column * width / pixelsWidth;
        var im = config.UpperLeft.Imaginary - row * height / pixelsHeight;
        var point = new Complex(re, im);

        var et = EscapeTime(point);
        pixels[row * pixelsWidth + column] = (!et.HasValue) ? 0 : 255 - et.Value;
      }
    }
  }

  static int? EscapeTime(Complex c)
  {
    var zr = 0.0;
    var zi = 0.0;

    for (int it = 0; it < MaxIterations; it++)
    {
      var zr2 = zr * zr;
      var zi2 = zi * zi;

      if (zr2 + zi2 > (EscapeRadius * EscapeRadius))
      {
        return it;
      }

      var temp = zr2 - zi2 + c.Real;
      zi = EscapeRadius * zr * zi + c.Imaginary;
      zr = temp;
    }
    return null;
  }
}